### PR TITLE
ALTS: Do not pass a host to grpc_channel_create_pollset_set_call(); let the channel supply it.

### PR DESCRIPTION
**Pull request by vandry**
When the handshaker_service_url is in "host:port" format such as it normally is when using ALTS in GCE (in which case it comes from then this makes no difference as the authority and the URL are the same. But when different URLs are used, the correct authority to use is not always the same as the URL. For example if the URL is unix:///some/path then the correct authority is "localhost". This is correctly computed by grpc_core::UnixResolverFactory and stored as the channel's default authority, but we throw that away when we override the authority for individual RPCs.

Note indeed that the majority of other callers of grpc_channel_create_* pass nullptr for the host/authority argument.

### CODE REVIEW

**Comment by markdroth** [comment]
This looks reasonable to me, but I'd like those more familiar with ALTS to verify that it won't cause any problems.  Reassigning this to @apolcyn.

**Comment by markdroth** [comment]
Should probably also be reviewed by @erm-g or someone on his team.

**Review by apolcyn — CHANGES_REQUESTED** [review]
This looks reasonable.

My only comment is about removing dead code.

Also still cc @erm-g, @matthewstevenson88 from security side

File: src/core/tsi/alts/handshaker/alts_handshaker_client.cc
Comment (matthewstevenson88) [reply src/core/tsi/alts/handshaker/alts_handshaker_client.cc:735]: It may be slightly tricky to remove the `handshaker_service_url` parameter since it's used on the next line:
```
 client->call = strcmp(handshaker_service_url, ALTS_HANDSHAKER_SERVICE_URL_FOR_TESTING) == 0 ? nullptr : ...
```
If this the `strcmp` is removed and `grpc_channel_create_pollset_set_call(...)` is always used instead, then we need to refactor the unit tests, which rely on this line, I think. If this is something you'd like to be done long-term, please file a bug and assign to me and I can clean it up when I have cycles.

**Comment by matthewstevenson88** [comment]
LGTM (modulo the minor dead code discussion). Thanks for the PR!

File: src/core/tsi/alts/handshaker/alts_handshaker_client.cc
Comment (erm-g) [reply src/core/tsi/alts/handshaker/alts_handshaker_client.cc:735]: @matthewstevenson88 IIUC `handshaker_service_url ` is not removed from the parameter list of `alts_grpc_handshaker_client_create` so it still can be used in `strcmp` LOC. The commit removes it only from `grpc_channel_create_pollset_set_call ` (replaces with `nullptr`) so `alts_handshaker_client.cc` looks fine.

However I couldn't find relevant tests across the codebase for `grpc_channel_create_pollset_set_call(...)` to make sure that authority will be calculated correctly. Could anybody point me in the right direction?

Comment (apolcyn) [reply src/core/tsi/alts/handshaker/alts_handshaker_client.cc:735]: One idea to test this:

We could start up the [fake ALTS handshaker service](https://github.com/grpc/grpc/blob/master/test/core/tsi/alts/fake_handshaker/fake_handshaker_server.cc) (an in-process gRPC server), and perform an ALTS handshake, setting the handshaker service url to something that needs this PR. We could add a field to the ALTS handshake server to check the RPC authority with [ExperimentalGetAuthority()](https://github.com/grpc/grpc/blob/c8407f8ef955ec064119790f41eb6e0469cc2df1/include/grpcpp/server_context.h#L300)

Comment (vandry) [reply src/core/tsi/alts/handshaker/alts_handshaker_client.cc:735]: As pointed out already, removing `handshaker_service_url` cannot be done without significantly refactoring the unit tests. IMHO such a refactor should not block this, so resolving. FWIW, I am not very familiar with the code but did also look for a good way to test that the correct authority is selected, but found nothing all that easy.

**Review by apolcyn — APPROVED** [review]
LGTM (discussions about dead code removal and testing don't need to block this PR)

**Review by matthewstevenson88 — APPROVED** [review]

**Pull request merged**
